## Exercici 1

Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe.  
Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants.  
Justifica la teva elecció.

In [2]:
import pandas as pd #esta libreria la utilizamos para cargar el dataframe
import numpy as np #esta libreria la utilizamos para reducir la dimensión y aplicar la función random de numpy para ver la muestra aleatoria

# cargamos el dataframe de iris.
df = pd.read_csv('DelayedFlights.csv')

# Reducimos la dimensión a 200.000 registros de manera aleatoria.
df_reduit = df.sample(n=200000, random_state=1)  # Utilizamos un valor de random_state para producirlo.

df_reduit

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1750755     6516830  2008     12          19          5   1332.0        1320   
1074983     3577214  2008      6          17          2   1731.0        1720   
46126        148177  2008      1          27          7    735.0         645   
327441      1035260  2008      2          13          3   1333.0        1230   
1363177     4599044  2008      8          12          2   1359.0        1346   
...             ...   ...    ...         ...        ...      ...         ...   
1595898     5763672  2008     10          27          1   2159.0        1920   
1503090     5275042  2008      9          25          4   2052.0        1930   
742602      2433686  2008      5          16          5   1335.0        1325   
1840527     6749307  2008     12          23          2    720.0         710   
1526456     5395875  2008     10           6          1   1657.0        1650   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1750755   1436.0        1430            WN  ...     7.0     8.0          0   
1074983   2013.0        1955            CO  ...     8.0    18.0          0   
46126      918.0         815            YV  ...     4.0    45.0          0   
327441    1515.0        1438            9E  ...     7.0    15.0          0   
1363177   1517.0        1447            FL  ...     5.0    15.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1595898    112.0        2225            MQ  ...     4.0    20.0          0   
1503090    115.0        2245            AA  ...    34.0    54.0          0   
742602    1422.0        1415            WN  ...     3.0     7.0          0   
1840527    839.0         831            DL  ...     9.0    20.0          0   
1526456   2217.0        2205            WN  ...     3.0    12.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1750755                 N         0           NaN           NaN      NaN   
1074983                 N         0          11.0           0.0      7.0   
46126                   N         0          63.0           0.0      0.0   
327441                  N         0           0.0           0.0      0.0   
1363177                 N         0          13.0           0.0     17.0   
...                   ...       ...           ...           ...      ...   
1595898                 N         0           0.0           0.0      8.0   
1503090                 N         0           0.0           0.0    150.0   
742602                  N         0           NaN           NaN      NaN   
1840527                 N         0           NaN           NaN      NaN   
1526456                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1750755           NaN                NaN  
1074983           0.0                0.0  
46126             0.0                0.0  
327441            0.0               37.0  
1363177           0.0                0.0  
...               ...                ...  
1595898           0.0              159.0  
1503090           0.0                0.0  
742602            NaN                NaN  
1840527           NaN                NaN  
1526456           NaN                NaN  

[200000 rows x 30 columns]

Con la función df.info nos muestra la tipología de datos que tiene cada columna, esto nos ayudará más adelante.

In [3]:
df.describe().T

count          mean           std     min        25%  \
Unnamed: 0         1936758.0  3.341651e+06  2.066065e+06     0.0  1517452.5   
Year               1936758.0  2.008000e+03  0.000000e+00  2008.0     2008.0   
Month              1936758.0  6.111106e+00  3.482546e+00     1.0        3.0   
DayofMonth         1936758.0  1.575347e+01  8.776272e+00     1.0        8.0   
DayOfWeek          1936758.0  3.984827e+00  1.995966e+00     1.0        2.0   
DepTime            1936758.0  1.518534e+03  4.504853e+02     1.0     1203.0   
CRSDepTime         1936758.0  1.467473e+03  4.247668e+02     0.0     1135.0   
ArrTime            1929648.0  1.610141e+03  5.481781e+02     1.0     1316.0   
CRSArrTime         1936758.0  1.634225e+03  4.646347e+02     0.0     1325.0   
FlightNum          1936758.0  2.184263e+03  1.944702e+03     1.0      610.0   
ActualElapsedTime  1928371.0  1.333059e+02  7.206007e+01    14.0       80.0   
CRSElapsedTime     1936560.0  1.343027e+02  7.134144e+01   -25.0       82.0   
AirTime            1928371.0  1.082771e+02  6.864261e+01     0.0       58.0   
ArrDelay           1928371.0  4.219988e+01  5.678472e+01  -109.0        9.0   
DepDelay           1936758.0  4.318518e+01  5.340250e+01     6.0       12.0   
Distance           1936758.0  7.656862e+02  5.744797e+02    11.0      338.0   
TaxiIn             1929648.0  6.812975e+00  5.273595e+00     0.0        4.0   
TaxiOut            1936303.0  1.823220e+01  1.433853e+01     0.0       10.0   
Cancelled          1936758.0  3.268348e-04  1.807562e-02     0.0        0.0   
Diverted           1936758.0  4.003598e-03  6.314722e-02     0.0        0.0   
CarrierDelay       1247488.0  1.917940e+01  4.354621e+01     0.0        0.0   
WeatherDelay       1247488.0  3.703571e+00  2.149290e+01     0.0        0.0   
NASDelay           1247488.0  1.502164e+01  3.383305e+01     0.0        0.0   
SecurityDelay      1247488.0  9.013714e-02  2.022714e+00     0.0        0.0   
LateAircraftDelay  1247488.0  2.529647e+01  4.205486e+01     0.0        0.0   

                         50%         75%        max  
Unnamed: 0         3242558.0  4972466.75  7009727.0  
Year                  2008.0     2008.00     2008.0  
Month                    6.0        9.00       12.0  
DayofMonth              16.0       23.00       31.0  
DayOfWeek                4.0        6.00        7.0  
DepTime               1545.0     1900.00     2400.0  
CRSDepTime            1510.0     1815.00     2359.0  
ArrTime               1715.0     2030.00     2400.0  
CRSArrTime            1705.0     2014.00     2400.0  
FlightNum             1543.0     3422.00     9742.0  
ActualElapsedTime      116.0      165.00     1114.0  
CRSElapsedTime         116.0      165.00      660.0  
AirTime                 90.0      137.00     1091.0  
ArrDelay                24.0       56.00     2461.0  
DepDelay                24.0       53.00     2467.0  
Distance               606.0      998.00     4962.0  
TaxiIn                   6.0        8.00      240.0  
TaxiOut                 14.0       21.00      422.0  
Cancelled                0.0        0.00        1.0  
Diverted                 0.0        0.00        1.0  
CarrierDelay             2.0       21.00     2436.0  
WeatherDelay             0.0        0.00     1352.0  
NASDelay                 2.0       15.00     1357.0  
SecurityDelay            0.0        0.00      392.0  
LateAircraftDelay        8.0       33.00     1316.0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

**Observaciones**: Observamos que en el caso de las columnas Cancelled y Diverted, dado que son variables binarias que indican si un vuelo fue cancelado o desviado (1) o no (0), sería más apropiado cambiar su formato a booleano para reflejar la naturaleza de la variable.

In [19]:
import pandas as pd

#llamamos al dataframe df entre corchetes le decimos la columna que queremos, y le decimos con la función .astype que queremos 
#cambiar el formato de esa columna a bool que es booleano.
df['Cancelled'] = df['Cancelled'].astype(bool)
df['Diverted'] = df['Diverted'].astype(bool)

#verificamos que se ha hecho el cambio correctamente
print(df.dtypes)

Unnamed: 0             int64
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled               bool
CancellationCode      object
Diverted                bool
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


En la muestra anterior, a simple vista observo que hay columnas con valores NaN:  

**CarrierDelay:** El tiempo de retraso debido a la aerolínea en minutos.  
**WeatherDelay:** El tiempo de retraso debido al clima en minutos.  
**NASDelay:** El tiempo de retraso debido al Sistema Nacional de Aviación en minutos.  
**SecurityDelay:** El tiempo de retraso debido a problemas de seguridad en minutos.  
**LateAircraftDelay:** El tiempo de retraso debido a problemas con el avión en minutos.  

Para poder comprobar si existen otras columnas con valores NaN, aplico el siguiente código.

In [5]:
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Al comprobar que exiten más columnas con valores NaN, me surgen varias duda:  
**¿Qué vuelos fueron cancelados, desvidados y retrasados que no hayan tenido relación con los motivos de retraso?**  Solución ejercicio 2. 

¿Cuántos vuelos fueron cancelados, desviados y retrasados?  
¿Los vuelos que han sido desviados han llegado con retraso?
¿Qué relación tienen las columnas de los motivos de retraso con el resto de columnas?  

**1. ¿Cuántos vuelos fueron cancelados, desviados y retrasados?**  

In [6]:
vuelos_cancelados = df['Cancelled'].sum()
vuelos_desviados = df['Diverted'].sum()
vuelos_retrasados = df[df['ArrDelay']>0].shape[0]
print('Vuelos cancelados: ',vuelos_cancelados)
print('Vuelos desviados: ', vuelos_desviados)
print('Vuelos retrasados: ', vuelos_retrasados)

Vuelos cancelados:  633
Vuelos desviados:  7754
Vuelos retrasados:  1723415


In [7]:
print('Total', vuelos_cancelados + vuelos_desviados + vuelos_retrasados)

Total 1731802


**¿Los vuelos que han sido desviados han llegado con retraso?**

In [8]:
vuelos_desviados_con_retraso = df[(df['Diverted'] == 1) & (df['ArrDelay']>0)].shape[0]
print(f'vuelos desviados que llegaron con retraso',{vuelos_desviados_con_retraso})

vuelos desviados que llegaron con retraso {0}


**Conclusión**: Podemos confirmar que los vuelos que fueron desviados **NO** tuvieron retraso.

**2. ¿Qué relación tienen las columnas de los motivos de retraso con las columnas retraso en la salida y retraso en la llegada?**  

Para ello vamos a utilizar el método .corr() de la librería pandas, que sirve para calcular la correlación entre columnas.  
Nos dará una matriz de correlación que muestra la relación entre pares de columnas.

**Valores cercanos a 1 indican una fuerte correlación positiva**, mientras que **valores cercanos a -1 incidcan una fuerte correlación negativa**. Quiero averiguar si las columnas de motivos (CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay) que tienen valores NaN, tienen relación con las columnas de cancelación, retrasos y/o desviación ( DepDelay, ArrDelay)

In [13]:
motivos_retrasos = ['CarrierDelay','WeatherDelay', 'NASDelay','SecurityDelay','LateAircraftDelay']
tiempo_retraso = ['DepDelay', 'ArrDelay']

motivos_retrasos_vs_tiempo_retraso = df[motivos_retrasos + tiempo_retraso].corr().loc[motivos_retrasos, tiempo_retraso]
print(motivos_retrasos_vs_tiempo_retraso)

                   DepDelay  ArrDelay
CarrierDelay       0.528038  0.501773
WeatherDelay       0.244488  0.264624
NASDelay           0.233976  0.387606
SecurityDelay     -0.002343 -0.002488
LateAircraftDelay  0.513041  0.478123


**Conclusión:**  Observamos que:
1. El tiempo de retraso debido a la aerolínea (CarrierDelay) tiene correlación con el tiempo de retraso en salida y en la llegada. 
2. El tiempo de retraso debido a problemas con el avión (LateAircraftDelay) tiene más correlación con el tiempo de retraso en la salida que con el tiempo de retraso en llegada (aunque ambos nos indica que tienen correlación).

El resto factores que indican el motivo del retraso NO tienen relación con el retraso en la salida y en la llegada.

In [81]:
#Como el dataframe es muy grande, lo vamos a reducir su tamaño a 200.000 registros

# Primero establecemos una semilla para reproducibilidad
import numpy as np
np.random.seed(42)

# Seleccionar un subconjunto aleatorio de 200,000 registros
n_df_muestreo = n_df.sample(n=200000)

# Verificar el nuevo tamaño del DataFrame
print(n_df_muestreo.shape)


(200000, 27)


## Exercici 2

Fes un informe complet del dataset:

Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
Troba quantes dades faltants hi ha per columna.
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Fes una taula de les aerolínies amb més endarreriments acumulats.
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
Aporta allò que consideris rellevant.

**1. Resumeix estadísticament el dataset i les columnes d’interès.**

In [83]:
n_df_muestreo.describe().T

count         mean          std        min          25%  \
DepTime            200000.0  1520.114805   450.087506   1.000000  1205.000000   
CRSDepTime         200000.0  1468.567445   424.431628   0.000000  1135.000000   
ArrTime            199267.0  1611.293159   548.219910   1.000000  1317.000000   
CRSArrTime         200000.0  1635.514525   464.072463   0.000000  1325.000000   
FlightNum          200000.0  2188.548195  1944.372086   1.000000   615.000000   
ActualElapsedTime  199127.0   133.222772    72.001365  15.000000    80.000000   
CRSElapsedTime     199974.0   134.184889    71.177993   1.000000    82.000000   
AirTime            199127.0   108.197070    68.472446   0.000000    58.000000   
ArrDelay           199127.0    42.371456    57.004644 -57.000000     9.000000   
DepDelay           200000.0    43.319360    53.581396   6.000000    12.000000   
Distance           200000.0   764.961740   572.856710  11.000000   340.000000   
TaxiIn             199267.0     6.817747     5.342789   0.000000     4.000000   
TaxiOut            199956.0    18.217083    14.410192   0.000000    10.000000   
CarrierDelay       128970.0    19.197666    44.126249   0.000000     0.000000   
WeatherDelay       128970.0     3.691052    20.863373   0.000000     0.000000   
NASDelay           128970.0    15.002566    33.913939   0.000000     0.000000   
SecurityDelay      128970.0     0.091223     2.014928   0.000000     0.000000   
LateAircraftDelay  128970.0    25.467403    42.275946   0.000000     0.000000   
AvgSpeed           199127.0          inf          NaN   0.354839     5.868852   
ArrivedLate        200000.0     0.890780     0.311916   0.000000     1.000000   

                           50%          75%     max  
DepTime            1546.000000  1901.000000  2400.0  
CRSDepTime         1510.000000  1817.000000  2359.0  
ArrTime            1716.000000  2031.000000  2400.0  
CRSArrTime         1707.000000  2015.000000  2400.0  
FlightNum          1549.000000  3423.000000  9740.0  
ActualElapsedTime   116.000000   165.000000   666.0  
CRSElapsedTime      116.000000   165.000000   660.0  
AirTime              90.000000   137.000000   644.0  
ArrDelay             25.000000    56.000000  2453.0  
DepDelay             24.000000    54.000000  2467.0  
Distance            607.000000   998.000000  4962.0  
TaxiIn                6.000000     8.000000   200.0  
TaxiOut              14.000000    21.000000   342.0  
CarrierDelay          2.000000    21.000000  2436.0  
WeatherDelay          0.000000     0.000000   868.0  
NASDelay              2.000000    15.000000  1030.0  
SecurityDelay         0.000000     0.000000   280.0  
LateAircraftDelay     8.000000    34.000000  1012.0  
AvgSpeed              6.729885     7.469848     inf  
ArrivedLate           1.000000     1.000000     1.0

**2. Fes una anàlisi estadístic del que consideris rellevant.** 

**¿Qué vuelos fueron cancelados, desvidados y retrasados que no hayan tenido relación con los motivos de retraso?** 

*este código nos indica lo mismo si quisieramos averiguar cuantos NaN tenemos en los vuelos con motivos.*

In [84]:
# Filtrar vuelos sin motivos de retraso que han sido cancelados, desviados o retrasados
vuelos_sin_motivos = n_df_muestreo[
    (n_df_muestreo['CarrierDelay'].isnull()) &
    (n_df_muestreo['WeatherDelay'].isnull()) &
    (n_df_muestreo['NASDelay'].isnull()) &
    (n_df_muestreo['SecurityDelay'].isnull()) &
    (n_df_muestreo['LateAircraftDelay'].isnull())
]

# Contar el número de vuelos cancelados sin motivos de retraso
num_cancelados_sin_motivos = vuelos_sin_motivos[vuelos_sin_motivos['Cancelled'] == 1].shape[0]

# Contar el número de vuelos desviados sin motivos de retraso
num_desviados_sin_motivos = vuelos_sin_motivos[vuelos_sin_motivos['Diverted'] == 1].shape[0]

# Contar el número de vuelos retrasados sin motivos de retraso
num_retrasados_sin_motivos = vuelos_sin_motivos[vuelos_sin_motivos['ArrDelay'] > 0].shape[0]

# Mostrar resultado
print("Número de vuelos cancelados sin motivos de retraso:", num_cancelados_sin_motivos)
print("Número de vuelos desviados sin motivos de retraso:", num_desviados_sin_motivos)
print("Número de vuelos retrasados sin motivos de retraso:", num_retrasados_sin_motivos)

print('Total', num_cancelados_sin_motivos + num_desviados_sin_motivos + num_retrasados_sin_motivos)


Número de vuelos cancelados sin motivos de retraso: 60
Número de vuelos desviados sin motivos de retraso: 813
Número de vuelos retrasados sin motivos de retraso: 49186
Total 50059


**Nota**: Sabemos que  
Total Vuelos cancelados:  633  
Total Vuelos desviados:  7754  
Total Vuelos retrasados:  1723415  
Total 1731802

y que 484.314 son los vuelos que han sido cancelados, desviados y retrasados sin motivos de retraso.

**Conclusiones**: Observamos que los vuelos cancelados y desviados son el mismo número y no tienen realación con los motivos. En cambio, los vuelos con retraso, tenemos 1.247.488 con motivo y 475.927 sin motivo. 

### ¿Que ha pasado con los 484314 que han sido cancelados, desviados y retrasados sin motivos?

Info:  
Número total de vuelos

In [88]:
# Obtener el número total de vuelos, utilizamos la función len que cuenta el total de filas del df.
#Como estamos cogiendo 200.000 registros, si ponemos el df original nos aparecen 
total_vuelos = len(df)

print("Número total de vuelos:", total_vuelos)



Número total de vuelos: 1936758


In [87]:
# Obtener el número total de vuelos, utilizamos la función len que cuenta el total de filas del df.
#Como estamos cogiendo 200.000 registros
total_vuelos = len(n_df_muestreo)

print("Número total de vuelos:", total_vuelos)


Número total de vuelos: 200000


¿Cuantos vuelos afectados con motivos tenemos?

In [89]:
# Primero vamos a averiguar la suma de vuelos con motivos de retraso
suma_vuelos_afectados_con_motivos = n_df_muestreo[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].gt(0).any(axis=1).sum()

print("Número de vuelos afectados con motivos de retraso:", suma_vuelos_afectados_con_motivos)


Número de vuelos afectados con motivos de retraso: 128970


¿Cuantos vuelos afectados tenemos?

In [90]:
# Segundo vamos a averiguar la suma de vuelos afectados, que incluyen los vuelos con motivos y los sin motivos.
vuelos_afectados = suma_vuelos_con_motivos + num_cancelados_sin_motivos + num_desviados_sin_motivos + num_retrasados_sin_motivos
vuelos_afectados

1297547

¿Cuántos vuelos sin afectaciones tenemos?

In [91]:
#Obtener el número de vuelos sin afectaciones
vuelos_sin_afectacion = total_vuelos - vuelos_afectados
vuelos_sin_afectacion

639211

### Lo cual quiere decir que de los 1.936.758 vuelos que tenemos en nuestro dataframe, 1.731.802 han sido afectados y 204.956 no.

**Ahora, teniendo en cuenta estos datos, sería interesante averiguar los o el motivo de los 484.314 vuelos afectados.
¿Cuales son las causas?**

1. Hipótesis:  
¿Es posible que tenga que ver con la distancia del vuelo (Distance)?

2. Hipótesis:  
¿Es posible que tenga que ver con la aerolinea?

3. Hipótesis:  
    ¿És posible que tenga que ver con la época del año?

.....

**4. Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)**

Velocidad media del vuelo

In [92]:
# Calcular velocidad media del vuelo (en millas por minuto) 'AvgSpeed' lo que estamos haciendo en dividir la distancia en millas por la hora 
#el tiempo en el aire en minutos.

n_df_muestreo['AvgSpeed'] = n_df_muestreo['Distance'] / n_df_muestreo['AirTime']

Si ha llegado tarde o no

In [94]:
# Creamos una columna que indica si ha llegado tarde (1 si sí, 0 si no)
n_df_muestreo['ArrivedLate'] = (n_df_muestreo['ArrDelay'] > 0).astype(int) #un vuelo llega tarde si ArrDelay es mayor que cero.

In [93]:
# Imprimir las primeras filas del DataFrame con las nuevas columnas seleccionadas
print(n_df_muestreo[['AirTime', 'Distance', 'AvgSpeed', 'ArrDelay', 'ArrivedLate']].head())


         AirTime  Distance  AvgSpeed  ArrDelay  ArrivedLate
1782417    177.0      1297  7.327684      71.0            1
512712      55.0       334  6.072727     548.0            1
447137      44.0       150  3.409091       9.0            1
55082       32.0       142  4.437500      52.0            1
877634     134.0       817  6.097015      22.0            1


**3. Troba quantes dades faltants hi ha per columna.**

In [95]:
n_df_muestreo.isnull().sum()

DepTime                  0
CRSDepTime               0
ArrTime                733
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime      873
CRSElapsedTime          26
AirTime                873
ArrDelay               873
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 733
TaxiOut                 44
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71030
WeatherDelay         71030
NASDelay             71030
SecurityDelay        71030
LateAircraftDelay    71030
AvgSpeed               873
ArrivedLate              0
dtype: int64

**5. Fes una taula de les aerolínies amb més endarreriments acumulats.**

In [96]:
# Hacemos un grupby para agrupar por aerolinea y sumar (.sum) los retrasos acumulados
endarreriments_per_aerolinia = n_df_muestreo.groupby('UniqueCarrier')['ArrDelay'].sum()

# Ordenamos las aerolineas según los atrasos acumulados de mayor a menor 
endarreriments_per_aerolinia = endarreriments_per_aerolinia.sort_values(ascending=False)

endarreriments_per_aerolinia

UniqueCarrier
WN    1178394.0
AA     910681.0
UA     696819.0
MQ     665562.0
OO     619404.0
XE     543941.0
DL     470912.0
CO     427113.0
EV     402521.0
YV     380899.0
US     379200.0
NW     344490.0
FL     323474.0
B6     315833.0
OH     272824.0
9E     245790.0
AS     146267.0
F9      83705.0
HA      27816.0
AQ       1656.0
Name: ArrDelay, dtype: float64

**6. Quins són els vols més llargs? I els més endarrerits?**

In [98]:
# Troba els vols més llargs
vols_mes_llargs = n_df_muestreo.loc[n_df_muestreo['Distance'].idxmax()]

#n_df.loc (localizame o ves) a la columna 'Distance'/ .idmax devuélveme el índice del vuelo con la distancia máxima.

print(vols_mes_llargs[['UniqueCarrier', 'FlightNum', 'Origin', 'Dest', 'Distance']])


UniqueCarrier      CO
FlightNum          15
Origin            EWR
Dest              HNL
Distance         4962
Name: 1620205, dtype: object


**7. Busca les rutes més llargues i les que acumulen més retards. Aporta allò que consideris rellevant.**

In [99]:
# Agrupamos por ruta (origen y destino) y sumamos las distancias y los retrasos
rutes_agrupades = n_df_muestreo.groupby(['Origin', 'Dest']).agg({'Distance': 'sum', 'ArrDelay': 'sum'}).reset_index()

# Para encontrar la ruta más larga aplicamos el mismo código que el punto anterior (6.)
ruta_mes_llarga = rutes_agrupades.loc[rutes_agrupades['Distance'].idxmax()]

# Para encontrar la ruta con más retrasos acumulados (idem)
ruta_amb_mes_endarreriments = rutes_agrupades.loc[rutes_agrupades['ArrDelay'].idxmax()]

# Imprimir los resultados
print("Ruta más larga:",ruta_mes_llarga)

print("\nRuta con más retrasos acumulados:", ruta_amb_mes_endarreriments)


Ruta más larga: Origin          SFO
Dest            JFK
Distance     659430
ArrDelay    10966.0
Name: 4268, dtype: object

Ruta con más retrasos acumulados: Origin          ORD
Dest            LGA
Distance     318122
ArrDelay    25121.0
Name: 3479, dtype: object


# Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

In [100]:
# Especifica la ruta y el nombre de archivo para guardar el DataFrame con las nuevas columnas
ruta_archivo_excel = 'C:\\Users\\vsanc\\OneDrive\\Escritorio\\CIBERNARIUM\\DataScience\\Sprint3\\n_dataset.xlsx'

# Guarda el DataFrame en un archivo Excel
n_df_muestreo.to_excel(ruta_archivo_excel, index=False)
